ymer/src/main/java/com/avanza/ymer/YmerSpaceSynchronizationEndpoint.java

```diff
@@ -48,71 +49,86 @@
 
 	private final MirroredObjectWriter mirroredObjectWriter;
 	private final ToggleableDocumentWriteExceptionHandler exceptionHandler;
-	private final PersistedInstanceIdRecalculationService persistedInstanceIdRecalculationService;
+	private final PersistedInstanceIdCalculationService persistedInstanceIdCalculationService;
 	private final SpaceMirrorContext spaceMirror;
 	private final ScheduledExecutorService scheduledExecutorService;
 	private final Set<ObjectName> registeredMbeans = new HashSet<>();
+	private final ReloadableYmerProperties ymerProperties;
 
+	private Integer currentNumberOfPartitions;
 	private ApplicationContext applicationContext;
 
-	public YmerSpaceSynchronizationEndpoint(SpaceMirrorContext spaceMirror) {
+	public YmerSpaceSynchronizationEndpoint(SpaceMirrorContext spaceMirror, ReloadableYmerProperties ymerProperties) {
 		exceptionHandler = ToggleableDocumentWriteExceptionHandler.create(
 				new RethrowsTransientDocumentWriteExceptionHandler(),
 				new CatchesAllDocumentWriteExceptionHandler());
 		this.spaceMirror = spaceMirror;
 		this.mirroredObjectWriter = new MirroredObjectWriter(spaceMirror, exceptionHandler);
-		this.persistedInstanceIdRecalculationService = new PersistedInstanceIdRecalculationService(spaceMirror);
+		this.persistedInstanceIdCalculationService = new PersistedInstanceIdCalculationService(spaceMirror, ymerProperties);
 		this.scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+		this.ymerProperties = ymerProperties;
+		this.currentNumberOfPartitions = GigaSpacesInstanceIdUtil.getNumberOfPartitionsFromSystemProperty().orElse(null);
 	}
 
 	@Override
 	public void onOperationsBatchSynchronization(OperationsBatchData batchData) {
-		mirroredObjectWriter.executeBulk(batchData);
+		mirroredObjectWriter.executeBulk(getInstanceMetadata(), batchData);
+	}
+
+	private InstanceMetadata getInstanceMetadata() {
+		return new InstanceMetadata(currentNumberOfPartitions, ymerProperties.getNextNumberOfInstances().orElse(null));
 	}
 
 	@Override
 	public void setApplicationContext(@Nonnull ApplicationContext applicationContext) throws BeansException {
 		this.applicationContext = applicationContext;
-		persistedInstanceIdRecalculationService.setApplicationContext(applicationContext);
+		persistedInstanceIdCalculationService.setApplicationContext(applicationContext);
 	}
 
 	@Override
 	public void onApplicationEvent(@Nonnull ContextRefreshedEvent event) {
 		if (event.getApplicationContext().equals(applicationContext)) {
-			schedulePersistedIdRecalculationIfNecessary();
+			// If number of partitions wasn't available as a system property, set number of partitions
+			// from space properties, which should be available once the space is started
+			if (currentNumberOfPartitions == null) {
```

Comment:
I think the space property should have precedence over the system property. Ie, the system property should only be used if the space property is not available.

Reply:
Sure, I'll change so that the space property takes precedence over the system property. 👍